### PR TITLE
r/sagemaker_endpoint_configuration - support `data_capture_config` + validations and test refactor

### LINKED ISSUE #14170: Sagemaker capture data feature
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Sagemaker has a useful feature called "capture data". It saves the requests and responses made to a Sagemaker endpoint into S3. This feature is exposed in the API.

### New or Affected Resource(s)

* aws_sagemaker_endpoint_configuration

### References

* https://docs.aws.amazon.com/sagemaker/latest/dg/model-monitor-data-capture.html
* https://awscli.amazonaws.com/v2/documentation/api/latest/reference/sagemaker/create-endpoint-config.html

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Supersedes #11484
Closes #14170
Relates #13826

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_sagemaker_endpoint_configuration -  support `data_capture_config`
resource_aws_sagemaker_endpoint_configuration -  add plan time validation for `production_variants.accelerator_type`, `production_variants.instance_type `
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerEndpointConfiguration_'
--- PASS: TestAccAWSSagemakerEndpointConfiguration_kmsKeyId (65.41s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_disappears (67.85s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_basic (68.95s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig (80.93s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight (84.77s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType (86.74s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_tags (140.43s)
```

### CODE REVIEW

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerEndpointConfiguration_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerEndpointConfiguration_ -timeout 120m
=== RUN   TestAccAWSSagemakerEndpointConfiguration_basic
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_basic
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== RUN   TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== RUN   TestAccAWSSagemakerEndpointConfiguration_tags
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_tags
=== RUN   TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== RUN   TestAccAWSSagemakerEndpointConfiguration_disappears
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_basic
=== CONT  TestAccAWSSagemakerEndpointConfiguration_tags
=== CONT  TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== CONT  TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType (50.42s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_disappears (58.39s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_basic (60.73s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight (62.53s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_kmsKeyId (65.66s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig (70.63s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_tags (85.68s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	85.739s
```

File: aws/resource_aws_sagemaker_endpoint_configuration_test.go [hunk 1]
```diff
@@ -252,17 +315,17 @@ func testAccCheckSagemakerEndpointConfigurationExists(n string) resource.TestChe
...
 	return fmt.Sprintf(`
-resource "aws_sagemaker_model" "foo" {
-  name               = %q
-  execution_role_arn = aws_iam_role.foo.arn
+resource "aws_sagemaker_model" "test" {
+  name               = %[1]q
+  execution_role_arn = aws_iam_role.test.arn
 
   primary_container {
     image = "174872318107.dkr.ecr.us-west-2.amazonaws.com/kmeans:1"
```
Comment (breathingdust) [inline aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: @DrFaust92 This container is only available in us-west-2 (so getting failures in GovCloud). Is there a way to look up a container based on the region for this test?

Comment (breathingdust) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: cc: @YakDriver

Comment (ewbankkit) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: Also fails outside the standard commercial test region:

```console
$ AWS_DEFAULT_REGION=us-east-2 make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerEndpointConfiguration_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerEndpointConfiguration_basic -timeout 120m
=== RUN   TestAccAWSSagemakerEndpointConfiguration_basic
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_basic
=== CONT  TestAccAWSSagemakerEndpointConfiguration_basic
    resource_aws_sagemaker_endpoint_configuration_test.go:64: Step 1/2 error: Error running apply: 2020/10/29 15:00:29 [DEBUG] Using modified User-Agent: Terraform/0.12.26 HashiCorp-terraform-exec/0.10.0
        
        Error: error creating Sagemaker model: ValidationException: Region of "174872318107.dkr.ecr.us-west-2.amazonaws.com/kmeans:1" does not match expected region of us-east-2. Cross region ECR image pulls are not allowed.
        	status code: 400, request id: 5b6a1f91-c444-461b-b71a-77cbdd48332d
        
        
--- FAIL: TestAccAWSSagemakerEndpointConfiguration_basic (127.61s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	127.660s
FAIL
GNUmakefile:27: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

Comment (ewbankkit) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: See https://docs.aws.amazon.com/sagemaker/latest/dg/sagemaker-algo-docker-registry-paths.html for the value in other regions.

Comment (ewbankkit) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: I foresee a new data source...

Comment (DrFaust92) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: yeah, I saw this and thought if I should create a map for that similar to other data source. but as its not the scope of the PR I didn't touch it. I think ill create a subsequent PR to handle this?

Comment (YakDriver) [reply aws/resource_aws_sagemaker_endpoint_configuration_test.go:323]: For short-term, what about a simple switch with hardcoded values? It's not ideal but neither is the Sagemaker API. Any subsequent data source would just be a bunch of hardcoded information.

```suggestion
    image = %[2]q
```

File: aws/resource_aws_sagemaker_endpoint_configuration_test.go [hunk 2]
```diff
@@ -252,17 +315,17 @@ func testAccCheckSagemakerEndpointConfigurationExists(n string) resource.TestChe
 
 func testAccSagemakerEndpointConfigurationConfig_Base(rName string) string {
```
Comment (YakDriver) [inline aws/resource_aws_sagemaker_endpoint_configuration_test.go:316]:
```suggestion
func testAccSagemakerEndpointConfigurationConfig_Base(rName string) string {
	imageID := ""
	switch testAccGetRegion() {
	case "us-west-2":
		imageID = "174872318107"
	case "us-west-1":
		imageID = "632365934929"
	case "us-east-1":
		imageID = "382416733822"
	case "us-gov-west-1":
		imageID = "226302683700"
	}
	registryPath = fmt.Sprintf("%s.dkr.ecr.%s.%s/kmeans:1", imageID, testAccGetRegion(), testAccGetPartitionDNSSuffix())
```

File: aws/resource_aws_sagemaker_endpoint_configuration_test.go [hunk 3]
```diff
@@ -277,17 +340,17 @@ data "aws_iam_policy_document" "assume_role" {
     }
   }
 }
-`, rName, rName)
+`, rName)
```
Comment (YakDriver) [inline aws/resource_aws_sagemaker_endpoint_configuration_test.go:343]:
```suggestion
`, rName, registryPath)
```

**Review by YakDriver — CHANGES_REQUESTED** [review]

**Comment by DrFaust92** [comment]
Thanks @YakDriver, added changes and tested :

```
--- PASS: TestAccAWSSagemakerEndpointConfiguration_disappears (72.06s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_kmsKeyId (74.59s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight (77.22s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_basic (80.55s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType (82.60s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig (101.76s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_tags (155.19s)
```

**Comment by DrFaust92** [comment]
@YakDriver, using the new datasource. tests are passing

**Comment by ewbankkit** [comment]
Verified acceptance test in default and non-default Commercial regions:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerEndpointConfiguration_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerEndpointConfiguration_ -timeout 120m
=== RUN   TestAccAWSSagemakerEndpointConfiguration_basic
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_basic
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== RUN   TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== RUN   TestAccAWSSagemakerEndpointConfiguration_tags
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_tags
=== RUN   TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== RUN   TestAccAWSSagemakerEndpointConfiguration_disappears
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_basic
=== CONT  TestAccAWSSagemakerEndpointConfiguration_tags
=== CONT  TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== CONT  TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType (52.42s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight (55.05s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_disappears (55.51s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig (57.98s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_basic (62.66s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_kmsKeyId (68.31s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_tags (85.88s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	85.945s
$ AWS_DEFAULT_REGION=us-east-2 make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerEndpointConfiguration_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerEndpointConfiguration_ -timeout 120m
=== RUN   TestAccAWSSagemakerEndpointConfiguration_basic
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_basic
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
=== RUN   TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== RUN   TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== RUN   TestAccAWSSagemakerEndpointConfiguration_tags
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_tags
=== RUN   TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== RUN   TestAccAWSSagemakerEndpointConfiguration_disappears
=== PAUSE TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_basic
=== CONT  TestAccAWSSagemakerEndpointConfiguration_tags
=== CONT  TestAccAWSSagemakerEndpointConfiguration_disappears
=== CONT  TestAccAWSSagemakerEndpointConfiguration_kmsKeyId
=== CONT  TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType
=== CONT  TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight
--- PASS: TestAccAWSSagemakerEndpointConfiguration_basic (35.44s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_InitialVariantWeight (40.34s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_dataCaptureConfig (45.77s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_disappears (52.71s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_productionVariants_AcceleratorType (53.40s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_kmsKeyId (57.36s)
--- PASS: TestAccAWSSagemakerEndpointConfiguration_tags (77.98s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	78.041s
```

**Review by YakDriver — APPROVED** [review]
I say we forge ahead and tackle the next level of GovCloud problems in subsequent PR(s).

**Pull request merged**